gnd/src/main/java/com/google/android/gnd/workers/TileDownloadWorker.java

```diff
@@ -32,116 +35,140 @@
-  private Result downloadTile(Tile tile) {
-    localDataStore
-        .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.IN_PROGRESS).build())
-        .blockingAwait();
-
-    return downloadTileFile(tile, Optional.empty());
-  }
-
-  /** Resumes downloading the source for {@param tile} marked as {@code Tile.State.IN_PROGRESS}. */
-  private Result resumeTileDownload(Tile tile) {
-    File existingTileFile = new File(context.getFilesDir(), tile.getPath());
-    HashMap<String, String> requestProperties = new HashMap<>();
-
-    requestProperties.put("Range", existingTileFile.length() + "-");
+  private Completable downloadTile(Tile tile) {
+    Map<String, String> requestProperties = new HashMap<>();
+
+    // To resume a download for an in progress tile, we use the HTTP Range request property.
+    // The range property takes a range of bytes, the server returns the content of the resource
+    // that corresponds to the given byte range.
+    //
+    // To resume a download, we get the current length, in bytes, of the file on disk.
+    // appending '-' to the byte value tells the server to return the range of bytes from the given
+    // byte value to the end of the file, e.g. '500-' returns contents starting at byte 500 to EOF.
+    //
+    // Note that length returns 0 when the file does not exist, so this correctly handles an edge
+    // case whereby the local DB has a tile state of IN_PROGRESS but none of the file has been
+    // downloaded yet (since then we'll fetch the range '0-', the entire file).
+    //
+    // For more info see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range
+    if (tile.getState() == State.IN_PROGRESS) {
+      File existingTileFile = new File(context.getFilesDir(), tile.getPath());
+      requestProperties.put("Range", "bytes=" + existingTileFile.length() + "-");
+    }
 
-    return downloadTileFile(tile, Optional.of(requestProperties));
+    return localDataStore
+        .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.IN_PROGRESS).build())
+        .andThen(
+            Completable.fromRunnable(
+                () -> {
+                  downloadTileFile(tile, requestProperties);
+                }))
+        .onErrorResumeNext(
+            e -> {
+              Timber.d(e, "Failed to download tile: %s", tile);
+              return localDataStore.insertOrUpdateTile(
+                  tile.toBuilder().setState(State.FAILED).build());
+            })
+        .andThen(
+            localDataStore.insertOrUpdateTile(tile.toBuilder().setState(State.DOWNLOADED).build()));
   }
 
   /**
    * Verifies that {@param tile} marked as {@code Tile.State.DOWNLOADED} in the local database still
    * exists in the app's storage. If the tile's source file isn't present, initiates a download of
    * source file.
    */
-  private Result checkDownload(Tile tile) {
+  private Completable checkDownload(Tile tile) {
```

Comment:
Naming nit: Since this both checking and downloads, maybe `downloadIfNotFound`? This can be done in a follow up though.

Reply:
Done! 6373a71